File: src/main/java/day2/PlayingWithInts.java
package main.java.day2;

import java.util.*;

/**
 * Created by Ionut on 31.03.2016.
 */
public class PlayingWithInts {

    public static void main(String[] args) {

        List<List<Integer>> data = Arrays.asList(
                Arrays.asList(1, 2, 3),
                Arrays.asList(3, 5),
                Arrays.asList(7, 3, 10),
                Arrays.asList(3, 5, 6)
        );

        // if isIntersection is `false` then we do union
        boolean isIntersection =  Math.random() > 0.5;

        Set<Integer> uniqueItems = IntsProcessor.aggregateInts_01(data, isIntersection);
        System.out.println(Arrays.toString(uniqueItems.toArray()));

    }
}

File: src/main/scala/day2/PlayingWithInts.scala
package scala.day2

/**
  * Created by Ionut on 31.03.2016.
  */
object PlayingWithInts {
  def main(args: Array[String]) {

    val data = Seq(
      Seq(1, 2, 3),
      Seq(3, 5),
      Seq(7, 3, 10),
      Seq(3, 5, 6)
    )

    // if isIntersection is `false` then we do union
    val isIntersection: Boolean = Math.random > 0.5


    // solution
    val aggregation = (first: Seq[Int], second: Seq[Int]) => if (isIntersection) first.intersect(second)
    else first.union(second)


    println(data.reduce(aggregation).toSet)

  }

}
